packages/flutter_tools/lib/src/android/adb.dart

```diff
@@ -0,0 +1,223 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:io';
+
+import '../base/logging.dart';
+import '../base/process.dart';
+
+// https://android.googlesource.com/platform/system/core/+/android-4.4_r1/adb/OVERVIEW.TXT
+// https://android.googlesource.com/platform/system/core/+/android-4.4_r1/adb/SERVICES.TXT
+
+/// A wrapper around the `adb` command-line tool and the adb server.
+class Adb {
+  static const int adbServerPort = 5037;
+
+  final String adbPath;
+
+  Adb(this.adbPath);
+
+  bool exists() {
+    try {
+      runCheckedSync([adbPath, 'version']);
+      return true;
+    } catch (_) {
+      return false;
+    }
+  }
+
+  /// Return the full text from `adb version`. E.g.,
+  ///
+  ///     Android Debug Bridge version 1.0.32
+  ///     Revision eac51f2bb6a8-android
+  ///
+  /// This method throws if `adb version` fails.
+  String getVersion() => runCheckedSync([adbPath, 'version']);
+
+  /// Starts the adb server. This will throw if there's an problem starting the
+  /// adb server.
+  void startServer() {
+    runCheckedSync([adbPath, 'start-server']);
+  }
+
+  /// Stops the adb server. This will throw if there's an problem stopping the
+  /// adb server.
+  void killServer() {
+    runCheckedSync([adbPath, 'kill-server']);
+  }
+
+  /// Ask the ADB server for its internal version number.
+  Future<String> getServerVersion() {
+    return _sendAdbServerCommand('host:version').then((String response) {
+      _AdbServerResponse adbResponse = new _AdbServerResponse(response);
+      if (adbResponse.isOkay)
+        return adbResponse.message;
+      throw adbResponse.message;
+    });
+  }
+
+  /// Queries the adb server for the list of connected adb devices.
+  Future<List<AdbDevice>> listDevices() async {
+    String stringResponse = await _sendAdbServerCommand('host:devices-l');
+    _AdbServerResponse response = new _AdbServerResponse(stringResponse);
+    if (response.isFail)
+      throw response.message;
+    String message = response.message.trim();
+    if (message.isEmpty)
+      return <AdbDevice>[];
+    return message.split('\n').map(
+      (String deviceInfo) => new AdbDevice(deviceInfo)).toList();
```

Comment:
nit: make the whitespace around delimiters symmetric. In this case you have a newline after the ( so you should have a newline before the matching ).